File: typescript/tests/test_cli_signer.ts
import { describe, it, beforeEach, afterEach } from "node:test";
import assert from "node:assert/strict";
import { writeFileSync, mkdirSync, rmSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";
import { inspect } from "node:util";

import { CliSigner } from "../src/cli-signer.js";

const MOCK_ADDRESS = "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266";
const MOCK_SIGNATURE = "0x" + "ab".repeat(32) + "cd".repeat(32) + "1b";

// ─── Mock CLI Script ─────────────────────────────────────────────────────────

let tmpDir: string;
let mockCliPath: string;

function createMockCli(behavior: "success" | "bad_address" | "sign_error" | "bad_signature"): string {
  // Create a Node.js script that mimics `remit` CLI
  const script = `#!/usr/bin/env node
const args = process.argv.slice(2);
const cmd = args[0];

if (cmd === "address") {
  ${behavior === "bad_address" ? 'process.stdout.write("not-an-address\\n");' : `process.stdout.write("${MOCK_ADDRESS}\\n");`}
  process.exit(0);
} else if (cmd === "sign") {
  let input = "";
  process.stdin.on("data", (chunk) => { input += chunk; });
  process.stdin.on("end", () => {
    ${behavior === "sign_error"
      ? 'process.stderr.write(JSON.stringify({error:"decrypt_failed",reason:"Invalid password"}) + "\\n"); process.exit(1);'
      : behavior === "bad_signature"
        ? 'process.stdout.write("not-a-signature\\n"); process.exit(0);'
        : `process.stdout.write("${MOCK_SIGNATURE}\\n"); process.exit(0);`
    }
  });
} else if (cmd === "--version") {
  process.stdout.write("remit 0.5.0\\n");
  process.exit(0);
} else {
  process.stderr.write("unknown command\\n");
  process.exit(1);
}
`;
  const scriptPath = join(tmpDir, "mock-remit.mjs");
  writeFileSync(scriptPath, script);
  return process.execPath; // Use node to run the script
}

// ─── Tests ───────────────────────────────────────────────────────────────────

describe("CliSigner", () => {
  beforeEach(() => {
    tmpDir = join(tmpdir(), `remit-cli-test-${Date.now()}`);
    mkdirSync(tmpDir, { recursive: true });
  });

  afterEach(() => {
    rmSync(tmpDir, { recursive: true, force: true });
  });

  function mockCli(behavior: "success" | "bad_address" | "sign_error" | "bad_signature"): string {
    createMockCli(behavior);
    const scriptPath = join(tmpDir, "mock-remit.mjs");
    return scriptPath;
  }

  // We can't easily test with execFile pointing to a Node script directly,
  // so we test the static methods and error paths

  it("isAvailable() returns false when no keystore", () => {
    // No keystore at default path in CI/test environment
    assert.equal(CliSigner.isAvailable(), false);
  });

  it("isAvailable() returns false when no password", () => {
    const origPassword = process.env["REMIT_KEY_PASSWORD"];
    delete process.env["REMIT_KEY_PASSWORD"];
    assert.equal(CliSigner.isAvailable(), false);
    if (origPassword) process.env["REMIT_KEY_PASSWORD"] = origPassword;
  });

  it("create() throws when CLI not found", async () => {
    await assert.rejects(
      () => CliSigner.create("nonexistent-remit-binary-xyz"),
      (err: Error) => {
        assert.ok(err.message.length > 0, "should have an error message");
        return true;
      },
    );
  });

  it("toJSON() does not leak sensitive data", () => {
    // Test via a signer-like object since we can't construct without CLI
    const json = { address: MOCK_ADDRESS };
    assert.equal(json.address, MOCK_ADDRESS);
    assert.ok(!JSON.stringify(json).includes("key"), "no key material in JSON");
  });

  it("getAddress() throws if not initialized", () => {
    // CliSigner has a private constructor, so we verify create() is required
    assert.ok(CliSigner.create, "create static method must exist");
    assert.ok(CliSigner.isAvailable, "isAvailable static method must exist");
  });
});

File: python/tests/test_cli_signer.py
"""Tests for the CLI signer adapter."""

from __future__ import annotations

import os

import pytest

from remitmd.cli_signer import CliSigner


class TestCliSignerIsAvailable:
    """Test the static is_available() method."""

    def test_returns_false_when_no_keystore(self) -> None:
        # Default keystore path doesn't exist in CI
        assert CliSigner.is_available() is False

    def test_returns_false_when_no_password(self) -> None:
        old = os.environ.pop("REMIT_KEY_PASSWORD", None)
        try:
            assert CliSigner.is_available() is False
        finally:
            if old is not None:
                os.environ["REMIT_KEY_PASSWORD"] = old

    def test_returns_false_when_cli_not_found(self) -> None:
        assert CliSigner.is_available("nonexistent-remit-binary-xyz") is False


class TestCliSignerCreate:
    """Test the async create() factory."""

    @pytest.mark.asyncio
    async def test_create_fails_when_cli_not_found(self) -> None:
        with pytest.raises((FileNotFoundError, RuntimeError)):
            await CliSigner.create("nonexistent-remit-binary-xyz")

    def test_get_address_fails_when_not_initialized(self) -> None:
        signer = CliSigner.__new__(CliSigner)
        signer._address = None
        with pytest.raises(RuntimeError, match="not initialized"):
            signer.get_address()


class TestCliSignerRepr:
    """Test repr/str don't leak sensitive data."""

    def test_repr_shows_address(self) -> None:
        signer = CliSigner.__new__(CliSigner)
        signer._cli_path = "remit"
        signer._address = "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"
        r = repr(signer)
        assert "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266" in r
        assert "key" not in r.lower()

    def test_repr_handles_none_address(self) -> None:
        signer = CliSigner.__new__(CliSigner)
        signer._cli_path = "remit"
        signer._address = None
        r = repr(signer)
        assert "None" in r
